rest/data/project_settings.go

```diff
@@ -97,6 +97,101 @@ func disableStartingSettings(p *model.ProjectRef) {
 	p.CommitQueue.Enabled = utility.FalsePtr()
 }
 
+// PromoteVarsToRepo moves variables from an attached project to its repo.
+// Promoted vars are removed from the project as part of this operation.
+// Variables whose names already appear in the repo settings will be overwritten.
+func PromoteVarsToRepo(projectId string, varNames []string, userId string) error {
+	project, err := model.GetProjectSettingsById(projectId, false)
+	if err != nil {
+		return errors.Wrapf(err, "getting project settings for project '%s'", projectId)
+	}
+
+	repoId := project.ProjectRef.RepoRefId
+
+	projectVars, err := model.FindOneProjectVars(projectId)
+	if err != nil {
+		return errors.Wrapf(err, "getting project variables for project '%s'", projectId)
+	}
+
+	repo, err := model.GetProjectSettingsById(repoId, true)
+	if err != nil {
+		return errors.Wrapf(err, "getting repo settings for repo '%s'", repoId)
+	}
+	repoVars, err := model.FindOneProjectVars(repoId)
+	if err != nil {
+		return errors.Wrapf(err, "getting repo variables for repo '%s'", repoId)
+	}
+
+	// Add each promoted variable to existing repo vars
+	apiRepoVars := &restModel.APIProjectVars{}
+	apiRepoVars.BuildFromService(*repoVars)
+	for _, varName := range varNames {
+		// Ignore nonexistent variables
+		if _, contains := projectVars.Vars[varName]; !contains {
+			continue
+		}
+		// Variables promoted from projects will overwrite matching repo variables
+		apiRepoVars.Vars[varName] = projectVars.Vars[varName]
+		if _, contains := projectVars.PrivateVars[varName]; contains {
+			apiRepoVars.PrivateVars[varName] = true
+		}
+		if _, contains := projectVars.AdminOnlyVars[varName]; contains {
+			apiRepoVars.AdminOnlyVars[varName] = true
+		}
+	}
+
+	if err = UpdateProjectVars(repoId, apiRepoVars, true); err != nil {
+		return errors.Wrapf(err, "adding variables from project '%s' to repo", projectId)
+	}
+
+	// Log repo update
+	repoAfter, err := model.GetProjectSettingsById(repoId, true)
+	if err != nil {
+		return errors.Wrapf(err, "getting settings for repo '%s' after adding promoted variables", repoId)
+	}
+	if err = model.LogProjectModified(repoId, userId, repo, repoAfter); err != nil {
+		return errors.Wrapf(err, "logging repo '%s' modified", repoId)
+	}
+
+	// Remove promoted variables from project
+	apiProjectVars := &restModel.APIProjectVars{
```

Comment:
Rather than doing this, can the vars list just be passed into the VarsToDelete APIProjectVars field?

Reply:
I tried this but `VarsToDelete` does not work in the way we both probably thought! It requires the ProjectVars struct to already be updated with the values removed, or else a `ConflictingUpdateOperators` is thrown due to the `$set` and `$unset` operations; this [test](https://github.com/evergreen-ci/evergreen/blob/b6792a6094208e2b4e522ad2ad170401d1180ab7/model/project_vars_test.go#L156-L177) is a good example of how it works.